### DISCUSSION TITLE
Cocotb types

### DISCUSSION THREAD

**Discussion by ktbarrett** [cocotb/cocotb#2612] [Proposals]
Moving discussion from [gitter](https://gitter.im/cocotb/Lobby?at=60e20a91457e19611a34cc83).

It's pretty well known that BinaryValue is broken, not just with incorrect behavior, but with an unintuitive interface. We have a chance and manifest to make something new to replace it. So what do you, the user, need out of the value types that are used with sim handles? And I don't mean just BinaryValue and logic array HDL types, but types for dealing with HDL arrays, records, strings, integers, etc. Is BinaryValue the only real problem?

Users have mentioned the following:
* Use `enum` values by name
* handle getting/setting record values
* assigning to signal handles directly with binary strings
* support for fixed point values

If you have a good idea or want to better expand on what you would like to see, feel free to post here!

**Comment by jwprice100** [comment]
Physical types (time) would also be nice. This limitation is truly easy to work around, but it would be convenient to not have too.

**Reply by alexforencich** [reply]
Some reworking of how time is handled would be nice.  Currently, I just do everything in terms of time steps as cocotb throws exceptions instead of rounding if I provide floating-point numbers.  This is for things like simulating packet serialization delays and similar; I have to compute the time delay value, but rounding the result is perfectly acceptable.

**Reply by eric-wieser** [reply]
Cocotb will behave if you use `Decimal` or `Fractions`s as your time units as those do not suffer from the imprecision of floats. That's not to say we don't need a time type, but it's probably a better interim solution than using steps directly.

**Reply by ktbarrett** [reply]
I think a Python time type (set up like a physical type) is a good idea. Ideally I'd like to integrate that into the scheduler so you can await on those objects like you can `Timer` (which would probably go away).

**Reply by eric-wieser** [reply]
I'm not overly keen on the idea of making arbitrary objects `await`able just for convenience, especially since I've not seen any other framework do this. Semantically a timer and a time are quite different things. So i think `await Timer(time)` should be here to stay; we'd just change `time` to be of type `Time` rather than a number.

**Reply by ktbarrett** [reply]
Then I think that `Timer` should be expanded to be that time type (really that's what I was suggesting all along). There really isn't a good reason to have two kinds of time types; one that supports `await`, and the other that supports useful operations, when one would do.

**Reply by alexforencich** [reply]
I'll take another look, but I'm not sure if using `Decimal` or `Fraction` would help in my case.  It's not really an issue with float imprecision.  

One of the computations in question looks like this:

    max(int(pkt.get_wire_size()*8*self.time_scale / (PCIE_GEN_RATE[self.cur_link_speed]*self.cur_link_width)), 1)

With reasonable values, the result before the cast to int might look like this:

    (16+256+8)*8*1e12 / (8e9*128/130 * 16) = 17773.4375

So even if I used something like `Decimal` instead of converting to steps myself, I would still get an exception here due to the truncation.  The exception would make more sense if it's changed to only occur if the result is truncated to 0.  

Incidentally, it would be nice to have a way to have a delay that's guaranteed to be at least one time step.  Basically, include `max(steps, 1)` internally or something along those lines.

**Reply by ktbarrett** [reply]
> include `max(steps, 1)` internally or something along those lines.

Since you can't wait half a step, you should wait the whole step. Instead of `round` one could use `ceil` to round waits to an integer number of steps. It's appropriate, and has the side effect of any non-0 infinitesimally small number becoming 1 step.

**Reply by alexforencich** [reply]
Sure, `max(ceil(steps), 1)` would be fine as well.  Keeping the `max` is probably a good idea to handle 0.

**Reply by eric-wieser** [reply]
I think some way to switch between "wait at least this much time"" and "wait exactly this much time" is needed (`Timer(..., exact=False)`?). The current behavior is useful as it lets you know when you're using times incompatible with your timestep - but sometimes you don't care about that.

**Comment by benbr8** [comment]
* Records: Could be done via dicts. Nested records/dicts should be a thing. Could be done for any named hierarchy subtree actually, not only records. 
* Fixed point: support for IEEE fixed point package notation with negative bit indices as fractional part would be great. I don't think BinaryValue is currently aware of its upper/lower index bounds though. Conversion to/from float with rounding/flooring would be great.
* HDL sign awareness: When reading negative signed values, the result should be negative. Simulators don't implement this consistently, so it would have to be handled by cocotb.
* The user should be able override any of this behaviour, for example if I want to interpret a bit vector as signed fixed with 3 bit fractional, that should be possible whether the HDL type matches or not.

These things go beyond the BinaryValue class and need additional info from the sim object it is tied to. Is there a reason for this separation? From the top of my head I can't think of a use case where I would need to use BinaryValue in isolation.

**Reply by eric-wieser** [reply]
> From the top of my head I can't think of a use case where I would need to use BinaryValue in isolation.

Here's one: Perform a transaction on a bus, and return a list containing the `BinaryValue` of that bus at every transaction. Those `BinaryValue`s will not be associated with a signal in any way.

**Reply by benbr8** [reply]
> Here's one: Perform a transaction on a bus, and return a list containing the `BinaryValue` of that bus at every transaction. Those `BinaryValue`s will not be associated with a signal in any way.

True, if it weren't broken you could use the value conversions `BinaryValue` provides later on. Otherwise one can just cast to a Python type and use that.

**Reply by eric-wieser** [reply]
If your bus has a bug and produces Xs on the first word, then converting to a python type will make your simulation crash immediately, and you won't learn that the rest of your words are OK. If you build the list of BinaryValues first, you can separately inspect your flow control and data pipelines, and pinpoint your issue faster.

**Reply by ktbarrett** [reply]
>  I don't think BinaryValue is currently aware of its upper/lower index bounds though

Right, this is a huge sore spot IMO.

> * The user should be able override any of this behaviour, for example if I want to interpret a bit vector as signed fixed with 3 bit fractional, that should be possible whether the HDL type matches or not.

In this case I was thinking conversion can be done with a pair of explicit casts, on the get() side by the user and on the set() side by the handle:

```python
>>> handle.value
LogicArray("011010", left=2, right=-3)
>>> s = Sfixed(handle.value)
>>> s
Sfixed(26, left=2, -3)
>>> float(s)
3.25
>>> int(s)
26
>>> str(s)
'011010'
>>> handle.value = s  # does LogicArray(s) cast on unlike types
```

`LogicArray` will have a cast delegate like `int` does using dunder methods. Something like `__logicarray__`. I'm not sure if changing the type by subclassing the handle type is particularly useful here. 

> Here's one: Perform a transaction on a bus, and return a list containing the BinaryValue of that bus at every transaction. Those BinaryValues will not be associated with a signal in any way.

Exactly, we need value types.

**Reply by eric-wieser** [reply]
@ktbarrett, I agree with everything in that example except `int(s)` and `str(s)`, which I think should either be an error or evaluate the fixed-point version of `int(float(s))` and `str(float(s))` - if you want to look at the underlying bits, you should be asking for them explicitly with `s.bits` or similar.

**Reply by alexforencich** [reply]
A design producing Xs is not necessarily a bug though.  Could be coming from a RAM that is not necessarily initialized yet.  For instance, if you write a couple of bytes to a RAM in a design and then read them out again, the bytes you didn't write could read as X, which is expected.  I have run in to this issue a few times before, and usually the simplest solution is to set COCOTB_RESOLVE_X.  At least cocotb is better than myhdl, where a signal with any x or z in it reads as 0.  What I would prefer to do is extract the segments in question.  If those segments contain Xs, then an exception would perhaps be warranted.  But due to deficiencies in BinaryValue, I extract the bits using normal python bit manipulation, which requires converting the whole signal to an integer first, and resolving the Xs at that time.  

Incidentally, how hard would it be to support "standard" bit manipulation operations on a cocotb object?  Bit shifts, bitwise AND, OR, XOR, invert, etc.  That would be super handy to be able to support that without having to resolve the Xs and Zs first.

**Reply by ktbarrett** [reply]
@alexforencich 

> how hard would it be to support "standard" bit manipulation operation?

Easy, the new types I've been developing support that.

>  set COCOTB_RESOLVE_X

That's a good topic of discussion. What should be done about that with the new system? I don't think such a feature should be associated with the value type. Preventing users from creating `BinaryValue`s with `X`s *at all* is a little ridiculous. If the new types support `X`s and `Z`s, do we even need `COCOTB_RESOLVE_X`, or is its utility of converting `X`s when reading from signals useful?

**Reply by alexforencich** [reply]
I think there needs to be a nicer way of handling it on a case-by-case basis.  The new types are nice, but at some point you will probably need to convert to integers, and at that time you'll need to either resolve the Xs or throw an exception.  COCOTB_RESOLVE_X is useful from a debugging and legacy compatibility standpoint, but what may make the most sense is to add some sort of resolve() method so the behavior can be selected on a case-by-case basis.  Not sure if that should be `<obj>.resolve()` (which might need to be special-cased for a method that could handle either integers or binary values), or `resolve(obj)` (which could fall-through for integers).  

I suppose another question is where things actually need to be resolved.  I use a lot of `bytes` and `bytearray`s.  When using the python builtins, things will have to be resolved first, which may not be optimal.  But, if there was an x/z-aware version of `bytes` and/or `bytearray`, then the resolution could be pushed even further down the line since it would not have to be done during re-packing into `bytes`/`bytearray`.
